File: Vachanam/Document/Parsers/DocumentParserResolver.swift
//
//  DocumentParserResolver.swift
//  Vachanam
//
//  Dispatches documents to the appropriate format parser based on file extension or source type.
//

import Foundation

public class DocumentParserResolver {
    public static let shared = DocumentParserResolver()
    
    public init() {}
    
    public func parse(source: DocumentSource, format: DocumentFormat? = nil) async throws -> ParsedDocument {
        let targetFormat: DocumentFormat
        if let fmt = format {
            targetFormat = fmt
        } else {
            switch source {
            case .fileURL(let url):
                targetFormat = DocumentFormat.detect(from: url)
            case .webURL:
                targetFormat = .webArticle
            case .rawText:
                targetFormat = .plainText
            }
        }
        
        switch targetFormat {
        case .pdf:
            throw DocumentParserError.invalidSource("PDF documents are handled by PDFKit / SentenceSegmenter")
        case .epub:
            return try await EPUBParser().parse(from: source)
        case .markdown:
            return try await MarkdownParser().parse(from: source)
        case .plainText:
            return try await PlainTextParser().parse(from: source)
        case .webArticle:
            return try await WebArticleParser().parse(from: source)
        }
    }
}

File: VachanamAndroid/app/src/main/java/com/vachanam/reader/data/parser/DocumentParserResolver.kt
package com.vachanam.reader.data.parser

import com.vachanam.reader.data.model.DocumentFormat
import java.io.File

class DocumentParserResolver {
    private val epubParser = EPUBParser()
    private val markdownParser = MarkdownParser()
    private val plainTextParser = PlainTextParser()
    private val webArticleParser = WebArticleParser()

    fun parserFor(format: DocumentFormat): DocumentParser {
        return when (format) {
            DocumentFormat.EPUB -> epubParser
            DocumentFormat.MARKDOWN -> markdownParser
            DocumentFormat.PLAIN_TEXT -> plainTextParser
            DocumentFormat.WEB_ARTICLE -> webArticleParser
            DocumentFormat.PDF -> plainTextParser // PDF handles its own extraction via PdfTextExtractor
        }
    }

    suspend fun parseFile(file: File): ParsedDocument {
        val format = DocumentFormat.detect(file)
        val parser = parserFor(format)
        return parser.parse(DocumentSource.LocalFile(file))
    }

    suspend fun parseUrl(url: String): ParsedDocument {
        val format = DocumentFormat.detectFromUrl(url)
        val parser = parserFor(format)
        return parser.parse(DocumentSource.WebUrl(url))
    }

    companion object {
        val shared = DocumentParserResolver()
    }
}
